### PR TITLE
Attach activation layer @open sesame 06/22 19:54

### LINKED ISSUE #153: Decouple activation to `layer`
Might be related to #152.
### AS-IS
Current implementation of `fc_layer` has activation function and loss function coupled.

### TO-BE
IMO, decoupling the activation layer as well as loss function(as @kparichay noted) is also needed to increase modularity.

for example in Keras, because activation functionality is decoupled, putting `bn_layer` between `fc` and `activation` can be done easily.

```python
model.add(Dense(64))
model.add(BatchNormalization())
model.add(Activation('tanh'))
```

still activation can be used as an argument of `fc_layer` possible because `fc_layer` can have `activation_layer`

```python
model.add(Dense(64, activation="relu" ))
```

### Proposal
- Add `activation_layer`(or equivalent) and extract the functionality.
- Change `fc_layer` to have `activation_layer`

@jijoongmoon Could you confirm if this change is okay?

### PR DESCRIPTION
**Pull request by zhoonit**
**Changes proposed in this PR:**
- strip activation function related members in `layer`
- init activation property as `activation_layer`

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped ~(conv::forward test need change)~

resolves #153 

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

### CODE REVIEW

File: nntrainer/include/layer.h [hunk 1]
```diff
@@ -36,8 +36,10 @@ namespace nntrainer {
  * @brief     Enumeration of cost(loss) function type
  *            0. MSR ( Mean Squared Roots )
  *            1. ENTROPY ( Cross Entropy )
- *            2. ENTROPY_SIGMOID (Cross Entropy amalgamated with sigmoid for stability)
- *            2. ENTROPY_SOFTMAX (Cross Entropy amalgamated with softmax for stability)
+ *            2. ENTROPY_SIGMOID (Cross Entropy amalgamated with sigmoid for
+ * stability)
+ *            2. ENTROPY_SOFTMAX (Cross Entropy amalgamated with softmax for
+ * stability)
  *            2. Unknown
```
Comment (jijoongmoon) [inline nntrainer/include/layer.h:43]: Numbers? is it related with this PR?

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 2]
```diff
@@ -495,6 +495,9 @@ TEST(nntrainer_Conv2DLayer, forwarding_01_p) {
   EXPECT_EQ(status, ML_ERROR_NONE);
   layer.setInputDimension(previous_dim);
 
+  nntrainer::ActivationLayer actLayer;
+  actLayer.setActivation(nntrainer::ACT_SIGMOID);
+
```
Comment (jijoongmoon) [inline test/unittest/unittest_nntrainer_layers.cpp:500]: This is not directly related with this PR, but we may need unit test case for the tizen api about activation layer.

File: nntrainer/include/layer.h (hunk 1 above)
Comment (zhoonit) [reply nntrainer/include/layer.h]: Not really related though, it seems it need fixing. I changed numbering :) Thanks.

File: test/unittest/unittest_nntrainer_layers.cpp (hunk 2 above)
Comment (zhoonit) [reply test/unittest/unittest_nntrainer_layers.cpp:500]: I think we need a discussion in #210. Thanks for pointing it out :)

File: nntrainer/src/fc_layer.cpp [hunk 3]
```diff
@@ -167,44 +160,10 @@ void FullyConnectedLayer::copy(std::shared_ptr<Layer> l) {
 }
 
 Tensor FullyConnectedLayer::backwarding(Tensor derivative, int iteration) {
-  Tensor djdb;
-  Tensor y2 = derivative;
-  Tensor y;
-
-  if (!last_layer) {
-    djdb = derivative.multiply(hidden.apply(activation_prime));
-  } else {
```
Comment (kparichay) [inline nntrainer/src/fc_layer.cpp:176]: I am not sure, but where does these cases go?

File: nntrainer/src/neuralnet.cpp [hunk 4]
```diff
@@ -442,17 +443,16 @@ int NeuralNetwork::initLossLayer() {
       ml_loge("Error: Cross Entropy not supported without softmax or sigmoid.");
       return ML_ERROR_NOT_SUPPORTED;
     }
```
Comment (zhoonit) [inline nntrainer/src/neuralnet.cpp:445]: The case has moved to here since network can evaluate if the last activation layer(if any) can be integrated into loss layer or just leave the way it is.

File: nntrainer/src/fc_layer.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/src/fc_layer.cpp:176]: Please refer to below comment :)

File: nntrainer/src/neuralnet.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:445]: How is `COST_MSR` handled here?

Comment (zhoonit) [reply nntrainer/src/neuralnet.cpp:445]: `COST_MSR` does not need to be handled here. Activation layer is added when `Network::init()` as long as user put activation props for the last layer .

The statement that `activation layer is last layer before adding loss layer` is not guaranteed though, it can be seen as user's fault. Do you think we need to warn the user if last layer is not activation layer before adding loss layer?

Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:445]: I meant that in case `fc_layer` is the last layer, gradient is handled differently for `COST_MSR` and `COST_ENTROPY`. I found that missing here.

Comment (zhoonit) [reply nntrainer/src/neuralnet.cpp:445]: This patch operates in this part as follows:

1. Add activation layer if `fc_layer` has activation prop.
2. in case of `COST_MSR`, just add `COST_MSR` loss layer
3. in case of `COST_ENTROPY` : 
   3-1. if last layer is not activation layer, throw error
   3-2. if last layer is sigmoid or softmax, **pop_back** that activation layer, and add `COST_ENTROPY_*` loss_layer.
   
```cpp
  if(!last_layer){
    djdb = derivative.multiply(hidden.apply(activation_prime));
 } switch (cost) {	
    case COST_MSR:	
      // this case does not need handling since activation layer(should have) added after last fc_layer (in #2).
      if (activation_type == ACT_SOFTMAX) {	
        y = hidden.apply(softmax);	
        djdb = derivative.multiply(hidden.apply(softmaxPrime));	
      } else {	
        djdb = derivative.multiply(hidden.apply(activation_prime));	
      }	
     
      break;	

    // since activation layer is removed before adding loss layer (#3-2) this case can be removed as well.
    case COST_ENTROPY_SIGMOID:	
      /** intended */	
    case COST_ENTROPY_SOFTMAX:	
      djdb = derivative;	
      break;	
    
    // shouldn't reach this part.
    case COST_ENTROPY:	
      throw std::runtime_error(	
        "Error: Cross Entropy not supported without softmax or sigmoid.");	
    case COST_UNKNOWN:	
      /** Intended */
```

revised

```cpp

 if (!last_layer) {
   djdb = derivative //.multiply(hidden.apply(activation) is done in activation layer 
 } else {
   switch (cost) {
    case COST_MSR:
      djdb = derivative; //.multiply(hidden.apply(...)) is done in activation layer
      break;
    case COST_ENTROPY_SIGMOID:
    case COST_ENTORPY_SOFTMAX:
      djdb = derivative;
      break;
    case OTHERS:
      //omitted
  } 
 }
```

So there is no need for a case handling. If my explanation is out of tune. I guess we can discuss this matter offline.

Out of topic though, I think there is a bug with this implementation.
Current implementation is `djdb = derivatives` so extra variable is not used. However, I think `djdb = derivatives.sum_by_batch()`. isn't it?

Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:445]: Further, applying activation prime for softmax and others is also different for COST_MSR.

Comment (zhoonit) [reply nntrainer/src/neuralnet.cpp:445]: > Further, applying activation prime for softmax and others is also different for COST_MSR.

It was just because before this patch `activation_prime` was `std::function<float (float)>` which cannot handle softmax.

Activation_layer.cpp handles them all together(checkout `activation_layer.cpp:76~122`)

Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:445]: Thanks for the explanation.
I understand how the `COST_MSR` is handled now. I think there is just 1 missing point about the softmax activation type.
For generic activation layers - 
```
djdb = derivative.multiply(hidden.apply(activation_prime));	
```
where hidden is the output of the fc/conv layer and activation is not yet applied.

However, for softmax
```
y = derivative.apply(softmax);
djdb = derivative.multiply(y.apply(activation_prime));
```
Attached an old version of this file https://github.com/nnstreamer/nntrainer/blob/2a650512813db6ce3bba828b5790066fbc655f14/nntrainer/src/fc_layer.cpp#L264-L267

I have added a PR #223 to reflect it back to the original version.
Please let me know if I understood it wrong.
